crates/core/src/operations/upsert.rs

```diff
@@ -315,69 +317,18 @@ impl UpsertBuilder {
-    ///
-    /// **Schema of returned DataFrame:**
-    /// - Each join key column (e\.g\. `"workspace_id"`, `"id"`)
-    /// - `__delta_rs_path` \(file path of the conflicting row\)
-    ///
-    /// **Example:**
-    /// If join keys are `["workspace_id", "id"]`, the resulting DataFrame schema:
-    /// ```
-    /// workspace_id: Int32
-    /// id: Utf8
-    /// __delta_rs_path: Utf8
-    /// ```
-    /// Each row represents a target record that conflicts with a source record \(i\.e\. same join key values\)\.
-    async fn find_conflicts(&self, target_df: &DataFrame) -> DeltaResult<DataFrame> {
-        let mut target_keys: Vec<_> = self.join_keys.iter().map(|k| col(k)).collect();
-        target_keys.push(col(FILE_PATH_COLUMN));
-
-        let source_keys: Vec<_> = self
-            .join_keys
-            .iter()
-            .map(|k| col(k).alias(&format!("source_{}", k)))
-            .collect();
-
-        let target_subset = target_df.clone().select(target_keys)?;
-        let source_subset = self.source.clone().select(source_keys)?;
-
-        let source_key_cols: Vec<_> = self
-            .join_keys
-            .iter()
-            .map(|s| format!("source_{}", s))
-            .collect();
-        let target_key_cols: Vec<_> = self.join_keys.iter().map(|s| format!("{}", s)).collect();
-
-        let conflicts = target_subset
-            .join(
-                source_subset,
-                JoinType::Inner,
-                &source_key_cols
-                    .iter()
-                    .map(|s| s.as_str())
-                    .collect::<Vec<&str>>(),
-                &target_key_cols
-                    .iter()
-                    .map(|s| s.as_str())
-                    .collect::<Vec<&str>>(),
-                None,
-            )?
-            //select only target columns (including file path)
-            .select(
-                target_key_cols
-                    .iter()
-                    .map(|s| col(s))
-                    .chain(std::iter::once(col(FILE_PATH_COLUMN)))
-                    .collect::<Vec<Expr>>(),
-            )
-            .map_err(|e| {
-                DeltaTableError::Generic(format!("Error during conflict detection join: {}", e))
-            });
-
-        conflicts
+    /// Prepare a DataFrame containing only the join key columns from the source.
+    /// This does not perform any matching or filtering against the target; it simply selects the relevant columns.
+    /// The resulting DataFrame is used later in an anti-join operation to filter out conflicting rows from the target DataFrame.
+    fn find_conflicts_keys_only(&self) -> DeltaResult<DataFrame> {
+        // Simply select join keys from source - we'll use this for the anti-join
+        let source_keys: Vec<_> = self.join_keys.iter().map(|k| col(k)).collect();
+        let source_subset =
+            self.source.clone().select(source_keys).map_err(|e| {
```

Comment:
The clone doesn't have an impact but implies that the same data may be queried twice. For this reason I like for methods on dataframes to consume the dataframe so it is very obvious at a higher level when when we are querying the same data again.